src/Http/Routing/src/Matching/DfaMatcherBuilder.cs

```diff
@@ -34,12 +36,26 @@ internal class DfaMatcherBuilder : MatcherBuilder
             ILoggerFactory loggerFactory,
             ParameterPolicyFactory parameterPolicyFactory,
             EndpointSelector selector,
-            IEnumerable<MatcherPolicy> policies)
+            IEnumerable<MatcherPolicy> policies,
+            bool? useLegacy30Behavior = null)
         {
             _loggerFactory = loggerFactory;
             _parameterPolicyFactory = parameterPolicyFactory;
             _selector = selector;
 
+            if (useLegacy30Behavior.HasValue)
+            {
+                _useLegacy30Behavior = useLegacy30Behavior.Value;
+            }
+            else if (AppContext.TryGetSwitch("Microsoft.AspNetCore.Routing.Use30CatchAllBehavior", out var enabled))
```

Comment:
Is it possible to add a unit test that tests setting the switch?

Reply:
Not really. https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/AppContext.cs#L19

It forces tests to be single-threaded since switches are global.

Reply:
Sounds like we need an IAppContextProvider

ahah... just kidding... unless?

Reply:
At some point you start staring into the void. The smaller the change the more it becomes tempting to unit test microdetails. I don't love it either, I wish that it was testable. 

We get into the problem **every** time we patch something that's worthy of an AppContext switch. Adding a unit test mitigates risk, making design changes to make things testable introduces risk.

Reply:
I was 90% [memeing](https://knowyourmeme.com/memes/just-kidding-unless) 😄 

We're always going to have flags like added in service releases. What about improving testability in 5.0? Pre-plan for 5.0.x

Reply:
The switch isn't present in 5.0. No one would prefer the broken behavior.

Reply:
I think what @JamesNK was implying was maybe we should do some general work in 5.0 to thread through something that would allow us to test appcontext switches.

Reply:
Yes. If we wait until we need it (servicing) then adding appcontext switch testing will always be too risky. We can plan ahead of time in 5.0 and have it ready for future 5.0.x.

I'll make an issue to discuss more.